### ISSUE TITLE
`deref_addrof` on raw pointer causes ` cannot take address of a temporary` error

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#14999]
Using the following flags

`--force-warn clippy::deref_addrof`

this code:

```rust
const A: *const i32 = &raw const *&2;
static B: () = { &raw const *&2; };
static mut C: *const i32 = &raw const *&2;

fn main() {}

```

caused the following diagnostics:

```
    Checking _main v0.1.0 (/tmp/icemaker_global_tempdir.8EVs4Ut5cNi4/icemaker_clippyfix_tempdir.nZ1QPnRWguZq/_main)
warning: immediately dereferencing a reference
 --> src/main.rs:1:34
  |
1 | const A: *const i32 = &raw const *&2;
  |                                  ^^^ help: try: `2`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#deref_addrof
  = note: requested on the command line with `--force-warn clippy::deref-addrof`

warning: immediately dereferencing a reference
 --> src/main.rs:2:29
  |
2 | static B: () = { &raw const *&2; };
  |                             ^^^ help: try: `2`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#deref_addrof

warning: immediately dereferencing a reference
 --> src/main.rs:3:39
  |
3 | static mut C: *const i32 = &raw const *&2;
  |                                       ^^^ help: try: `2`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#deref_addrof

warning: `_main` (bin "_main") generated 3 warnings (run `cargo clippy --fix --bin "_main"` to apply 3 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s

```

However after applying these diagnostics, the resulting code:

```rust
const A: *const i32 = &raw const 2;
static B: () = { &raw const 2; };
static mut C: *const i32 = &raw const 2;

fn main() {}

```
no longer compiled:

```
    Checking _main v0.1.0 (/tmp/icemaker_global_tempdir.8EVs4Ut5cNi4/icemaker_clippyfix_tempdir.nZ1QPnRWguZq/_main)
error[E0745]: cannot take address of a temporary
 --> src/main.rs:1:34
  |
1 | const A: *const i32 = &raw const 2;
  |                                  ^ temporary value

error[E0745]: cannot take address of a temporary
 --> src/main.rs:2:29
  |
2 | static B: () = { &raw const 2; };
  |                             ^ temporary value

error[E0745]: cannot take address of a temporary
 --> src/main.rs:3:39
  |
3 | static mut C: *const i32 = &raw const 2;
  |                                       ^ temporary value

For more information about this error, try `rustc --explain E0745`.
error: could not compile `_main` (bin "_main") due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `_main` (bin "_main" test) due to 3 previous errors

```

Version:
```
rustc 1.89.0-nightly (cf423712b 2025-06-05)
binary: rustc
commit-hash: cf423712b9e95e9f6ec84b1ecb3d125e55ac8d56
commit-date: 2025-06-05
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5

```
